typescript/token/src/adapters/EvmTokenAdapter.ts

```diff
@@ -0,0 +1,213 @@
+    public readonly contractFactory: any = ERC20__factory,
+  ) {
+    super(chainName, multiProvider);
+    this.contract = contractFactory.connect(
+      addresses.token,
+      this.getProvider(),
+    );
+  }
+
+  override async getBalance(address: Address): Promise<string> {
+    const balance = await this.contract.balanceOf(address);
+    return balance.toString();
+  }
+
+  override async getMetadata(isNft?: boolean): Promise<MinimalTokenMetadata> {
+    const [decimals, symbol, name] = await Promise.all([
+      !isNft ? this.contract.decimals() : 0,
+      this.contract.symbol(),
+      this.contract.name(),
+    ]);
+    return { decimals, symbol, name };
+  }
+
+  override populateApproveTx({
+    weiAmountOrId,
+    recipient,
+  }: TransferParams): Promise<PopulatedTransaction> {
+    return this.contract.populateTransaction.approve(recipient, weiAmountOrId);
+  }
+
+  override populateTransferTx({
+    weiAmountOrId,
+    recipient,
+  }: TransferParams): Promise<PopulatedTransaction> {
+    return this.contract.populateTransaction.transfer(recipient, weiAmountOrId);
+  }
+}
+
+// Interacts with Hyp Synthetic token contracts (aka 'HypTokens')
+export class EvmHypSyntheticAdapter<T extends HypERC20 = HypERC20>
+  extends EvmTokenAdapter<T>
+  implements IHypTokenAdapter
+{
+  constructor(
+    public readonly chainName: ChainName,
+    public readonly multiProvider: MultiProtocolProvider,
+    public readonly addresses: { token: Address },
+    public readonly contractFactory: any = HypERC20__factory,
+  ) {
+    super(chainName, multiProvider, addresses, contractFactory);
+  }
+
+  getDomains(): Promise<Domain[]> {
+    return this.contract.domains();
+  }
+
+  async getRouterAddress(domain: Domain): Promise<Buffer> {
+    const routerAddressesAsBytes32 = await this.contract.routers(domain);
+    // Evm addresses will be padded with 12 bytes
+    if (routerAddressesAsBytes32.startsWith('0x000000000000000000000000')) {
+      return Buffer.from(
+        strip0x(bytes32ToAddress(routerAddressesAsBytes32)),
+        'hex',
+      );
+      // Otherwise leave the address unchanged
+    } else {
+      return Buffer.from(strip0x(routerAddressesAsBytes32), 'hex');
+    }
+  }
+
+  async getAllRouters(): Promise<Array<{ domain: Domain; address: Buffer }>> {
+    const domains = await this.getDomains();
+    const routers: Buffer[] = await Promise.all(
+      domains.map((d) => this.getRouterAddress(d)),
+    );
+    return domains.map((d, i) => ({ domain: d, address: routers[i] }));
+  }
+
+  async quoteGasPayment(destination: Domain): Promise<string> {
+    const gasPayment = await this.contract.quoteGasPayment(destination);
+    return gasPayment.toString();
+  }
+
+  populateTransferRemoteTx({
+    weiAmountOrId,
+    destination,
+    recipient,
+    txValue,
+  }: TransferRemoteParams): Promise<PopulatedTransaction> {
+    const recipBytes32 = addressToBytes32(addressToByteHexString(recipient));
```

Comment:
recipient isn't bytes32 already?

Reply:
No, recipient would look like a normal address going in here. An unpadded hex string